### PR TITLE
Reference mapping with MuData-based models

### PR DESCRIPTION
**Pull request by ethanweinberger**
(Addresses #2393) While certain multimodal models (e.g. totalVI) can be used with either `AnnData` or `MuData` objects, currently the `ArchesMixin` class only supports `AnnData`-based models and so reference mapping cannot be performed with `MuData`-based models. This PR thus extends `ArchesMixin` class to handle MuData-based models by making appropriate changes to the `load_query_data` function and adding a new `prepare_query_mudata` function.

@martinkim0 my initial PR is still a work-in-progress (e.g. no unit tests yet), but I thought it would be helpful to get some initial feedback from you. To avoid duplicating code, my instinct for the `prepare_query_mudata` function was to essentially be a lightweight wrapper that calls `prepare_query_anndata` for each modality in the `MuData` object. However, when attempting to do this naively I quickly ran into issues with `var_names` for `MuData` objects being saved as a single list for all modalities. In particular, because variable names don't have identifiers for their modality of origin, I ran into issues where e.g. a protein-modality-`AnnData` would erroneously be padded with zeros for "missing" RNA-modality features. To work around this, I modified the base `save` function such that variable names are saved as a dictionary (with modalities as keys) for `MuData` objects.

This worked (as in, my code ran without errors and current tests still pass), but I imagine this would lead to backwards incompatibility issues for previously trained `MuData`-based models. Moreover, I still needed a way to tell `setup_anndata` which specific variable names to consider ahead of time, which seemed a bit ugly to me. Do you have ideas for a cleaner way to associate features with modalities?

### CODE REVIEW

**Comment by martinkim0** [comment]
Thanks for taking a stab at this! You're right - it looks like we should save var_names separately for each modality, so it's unfortunate that we don't already do this. 

This would actually be a great feature to include in our v2 release, since our plan is to rework large parts of the codebase. So there would be no issue with backwards compatibility. 

The other alternative is to use this new save method from now on, with the caveat that we'd have to implement a method to convert "legacy" MuData saves (similar to [`convert_legacy_save`](https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.base.BaseModelClass.html#scvi.model.base.BaseModelClass.convert_legacy_save)). 

I'm personally leaning more towards the first option since there would be less room for user confusion and maintenance issues. Let me know what you think! Is this a feature you're looking to use in a new model?

**Review by martinkim0 — COMMENTED** [review]
Thanks for taking a stab at this! You're right - it looks like we should save var_names separately for each modality, so it's unfortunate that we don't already do this. 

This would actually be a great feature to include in our v2 release, since our plan is to rework large parts of the codebase. So there would be no issue with backwards compatibility. 

The other alternative is to use this new save method from now on, with the caveat that we'd have to implement a method to convert "legacy" MuData saves (similar to [`convert_legacy_save`](https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.base.BaseModelClass.html#scvi.model.base.BaseModelClass.convert_legacy_save)). 

I'm personally leaning more towards the first option since there would be less room for user confusion and maintenance issues. Let me know what you think! Is this a feature you're looking to use in a new model?

File: scvi/model/base/_archesmixin.py [hunk 1]
```diff
@@ -8,11 +8,13 @@
 import pandas as pd
 import torch
 from anndata import AnnData
+from muon import MuData
```
Comment (martinkim0) [inline scvi/model/base/_archesmixin.py:11]: nit: need to import from `mudata` since we don't directly depend on `muon`

**Comment by ethanweinberger** [comment]
@martinkim0 with the closing of #2769, I've gone ahead and taken another stab at this PR.

The biggest change now compared to main is that I factored out most of the old code in `prepare_query_anndata` into a new function `_pad_and_sort_query_anndata` that's reused across `prepare_query_anndata` and a new `prepare_query_mudata` method.

Let me know what you think/if there's anything else you'd like in this PR in terms of testing/documentation/etc.

**Comment by ethanweinberger** [comment]
Hi @martinkim0. Just wanted to check in to see if you had any feedback for the latest version of this PR. Thanks!

File: src/scvi/model/base/_archesmixin.py [hunk 2]
```diff
@@ -326,7 +359,68 @@ def _get_loaded_data(reference_model, device=None):
...
+            var_names = reference_model.adata.var_names
         load_state_dict = deepcopy(reference_model.module.state_dict())
 
     return attr_dict, var_names, load_state_dict
+
+
+def _pad_and_sort_query_anndata(
+    adata,
+    reference_var_names,
+    inplace,
```
Comment (martinkim0) [inline src/scvi/model/base/_archesmixin.py:377]: typing here would be great!

Comment (martinkim0) [inline src/scvi/model/base/_archesmixin.py:377]: typing here would be great!

File: src/scvi/model/base/_archesmixin.py [hunk 3]
```diff
@@ -326,7 +359,68 @@ def _get_loaded_data(reference_model, device=None):
...
         attr_dict = reference_model._get_user_attributes()
         attr_dict = {a[0]: a[1] for a in attr_dict if a[0][-1] == "_"}
-        var_names = reference_model.adata.var_names
+        if isinstance(reference_model.adata, MuData):
+            var_names = {
+                mod: reference_model.adata[mod].var_names
+                for mod in reference_model.adata.mod.keys()
+            }
+        else:
+            var_names = reference_model.adata.var_names
```
Comment (martinkim0) [inline src/scvi/model/base/_archesmixin.py:368]: this could be replaced by a call to `scvi.model.base._save_load._get_var_names`

**Review by martinkim0 — COMMENTED** [review]
Thanks a lot! This looks great - I think refactoring out `_pad_and_sort_query_anndata` makes a lot of sense. Just left some minor comments regarding typing and getting var names. 

One last thing: could you add a test for `prepare_query_mudata` as well as a release note in `CHANGELOG.md` (this can go under the v1.2 release).

File: src/scvi/model/base/_archesmixin.py (hunk 3 above)
Comment (ethanweinberger) [reply src/scvi/model/base/_archesmixin.py]: Done. Thanks for the pointer!

File: src/scvi/model/base/_archesmixin.py (hunk 2 above)
Comment (ethanweinberger) [reply src/scvi/model/base/_archesmixin.py]: Done

**Comment by ethanweinberger** [comment]
> Thanks a lot! This looks great - I think refactoring out `_pad_and_sort_query_anndata` makes a lot of sense. Just left some minor comments regarding typing and getting var names.
> 
> One last thing: could you add a test for `prepare_query_mudata` as well as a release note in `CHANGELOG.md` (this can go under the v1.2 release).

Added the requested typing + incorporated your suggestion on `._get_var_names`. For testing I added a `test_scarches_mudata_prep_layer` in `test_totalvi.py` analogous to `test_scarches_data_prep_layer` in `test_scvi.py`.

**Review by martinkim0 — APPROVED** [review]
LGTM! thanks a lot for addressing the comments

File: CHANGELOG.md [hunk 4]
```diff
@@ -40,6 +40,7 @@ to [Semantic Versioning]. Full commit history is available in the
     passing in `gene_likelihood="normal"` {pr}`2780`.
 - Add {class}`scvi.external.MRVI` for modeling sample-level heterogeneity in single-cell RNA-seq
     data {pr}`2756`.
+- Add support for reference mapping with `MuData` models to {class}`scvi.model.base.ArchesMixin` {pr}`2578`.
```
Comment (martinkim0) [inline CHANGELOG.md:43]:
```suggestion
- Add support for reference mapping with {class}`mudata.MuData` models to {class}`scvi.model.base.ArchesMixin` {pr}`2578`.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,6 +40,7 @@ to [Semantic Versioning]. Full commit history is available in the
     passing in `gene_likelihood="normal"` {pr}`2780`.
 - Add {class}`scvi.external.MRVI` for modeling sample-level heterogeneity in single-cell RNA-seq
     data {pr}`2756`.
+- Add support for reference mapping with {class}`mudata.MuData` models to {class}`scvi.model.base.ArchesMixin` {pr}`2578`.
```
Comment (martinkim0) [inline CHANGELOG.md:43]:
```suggestion
- Add support for reference mapping with {class}`mudata.MuData` models to
    {class}`scvi.model.base.ArchesMixin` {pr}`2578`.
```

**Pull request merged**